app/src/org/commcare/android/resource/AndroidResourceManager.java

```diff
@@ -0,0 +1,258 @@
+     */
+    private void loadProfileViaTemp(Resource upgradeProfile)
+            throws UnfullfilledRequirementsException,
+            UnresolvedResourceException,
+            InstallCancelledException {
+        tempUpgradeTable.destroy();
+        loadProfileIntoTable(tempUpgradeTable, profileRef);
+        Resource tempProfile =
+                tempUpgradeTable.getResourceWithId(CommCarePlatform.APP_PROFILE_RESOURCE_ID);
+
+        if (tempProfile != null && tempProfile.isNewer(upgradeProfile)) {
+            upgradeTable.destroy();
+            tempUpgradeTable.copyToTable(upgradeTable);
+        }
+
+        tempUpgradeTable.destroy();
+    }
+
+    /**
+     * Set listeners and checkers that enable communication between low-level
+     * resource installation and top-level app update/installation process.
+     *
+     * @param tableListener  allows resource table to report its progress to the
+     *                       launching process
+     * @param cancelCheckker allows resource installers to check if the
+     *                       launching process was cancelled
+     */
+    @Override
+    public void setUpgradeListeners(TableStateListener tableListener,
+                                    InstallCancelled cancelCheckker) {
+        super.setUpgradeListeners(tableListener, cancelCheckker);
+
+        tempUpgradeTable.setStateListener(tableListener);
+        tempUpgradeTable.setInstallCancellationChecker(cancelCheckker);
+    }
+
+    /**
+     * Save upgrade stats if the upgrade was cancelled and wasn't complete at
+     * that time.
+     */
+    public void upgradeCancelled() {
+        if (!isUpgradeTableStaged()) {
+            UpdateStats.saveStatsPersistently(app, updateStats);
+        } else {
+            Log.i(TAG, "Upgrade cancelled, but already finished with these stats");
+            Log.i(TAG, updateStats.toString());
+        }
+    }
+
+    public void incrementUpdateAttempts() {
+        updateStats.registerStagingAttempt();
+    }
+
+    /**
+     * Log update failure that occurs while trying to install the staged update table
+     */
+    public void recordUpdateInstallFailure(Exception exception) {
+        updateStats.registerUpdateException(exception);
+    }
+
+    public void recordUpdateInstallFailure(AppInstallStatus result) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+    }
+
+    /**
+     * Clear update table, log failure with update stats,
+     * and, if appropriate, schedule a update retry
+     *
+     * @param result update attempt result
+     * @param ctx    Used for showing pinned notification of update task retry
+     */
+    public void processUpdateFailure(AppInstallStatus result, Context ctx) {
+        updateStats.registerUpdateException(new Exception(result.toString()));
+
+        boolean reusePartialTable =
+                (result == AppInstallStatus.UnknownFailure ||
```

Comment:
One thing you may find helpful here is that enumerations can have properties set on them. You could set the AppInstallStatus enumeration to have "can reuse in future" value that can be applied to new enumeration outcomes, which would be easier to maintain than have to track back to here.